pkg/controller/gitsourceanalysis/gitsourceanalysis_controller.go

```diff
@@ -0,0 +1,155 @@
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileGitSourceAnalysis{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("gitsourceanalysis-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource GitSourceAnalysis
+	err = c.Watch(&source.Kind{Type: &v1alpha1.GitSourceAnalysis{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileGitSourceAnalysis{}
+
+// ReconcileGitSourceAnalysis reconciles a GitSourceAnalysis object
+type ReconcileGitSourceAnalysis struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a GitSourceAnalysis object and makes changes based on the state read
+// and what is in the GitSourceAnalysis.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileGitSourceAnalysis) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := controllerLogger.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling GitSourceAnalysis")
+
+	// Fetch the GitSourceAnalysis instance
+	gsAnalysis := &v1alpha1.GitSourceAnalysis{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, gsAnalysis)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "There was an error while reading the GitSourceAnalysis object")
+		return reconcile.Result{}, err
+	}
+
+	if gsAnalysis.Status.Analyzed {
+		reqLogger.WithValues("git-source", gsAnalysis.Spec.GitSourceRef).
+			Info("Skipping GitSourceAnalysis as it was already analyzed")
+		return reconcile.Result{}, err
+	}
+
+	return analyze(reqLogger, r.client, gsAnalysis, request.Namespace)
+}
+
+func analyze(logger logr.Logger, client client.Client, gsAnalysis *v1alpha1.GitSourceAnalysis, namespace string) (reconcile.Result, error) {
+	gitSource := &v1alpha1.GitSource{}
+	err := client.Get(context.TODO(), newNsdName(namespace, gsAnalysis.Spec.GitSourceRef.Name), gitSource)
+	if err != nil {
```

Comment:
If we reach this condition `err != nil`,
then there should be a return statement also `return nil, err`, right?

Reply:
why? the error msg is set to the GitSourceAnalysis CR that is then updated. If the error was returned, then the reconciler would be triggered again which is something that we don't want